### ISSUE TITLE
Expected modifiers property to be always an integer value

### ISSUE THREAD

**Issue by krzysiekpiasecki** [FriendsOfPHP/Sami#248]
File: ClassReflection.php
---

Expected ```$this->modifiers``` to be an integer value but is non-numeric string

```php
 public function isAbstract()
 {
        return self::MODIFIER_ABSTRACT === (self::MODIFIER_ABSTRACT & $this->modifiers);
 }
```

Used dependencies:
---

```json
"require": {
    "php": "^7.1"
  },
  "require-dev": {
    "sami/sami": "^3.3",
  },
```

Error
---

This causes warning and finally twig runtime exception when updating a documentation. 

>> [Twig_Error_Runtime]                                                                                     
  An exception has been thrown during the rendering of a template ("Warning: A non-numeric value encounte  
  red in /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Reflection/ClassReflecti  
  on.php line 258") in "class.twig" at line 74.                

> [ErrorException]                                                                                         
  Warning: A non-numeric value encountered in /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sa  
  mi/sami/Sami/Reflection/ClassReflection.php line 258

**Comment by aik099** [comment]
Can you please:

1. dump the value of `$this->modifiers` via `var_dump`
2. specify PHP file, that causes this error to happen during parsing
3. show exact versions of composer dependencies (output of `composer show -i` command)

**Comment by krzysiekpiasecki** [comment]
You should be able to reproduce bug with:
```
git clone https://github.com/krzysiekpiasecki/Bowling.git
composer install
```

There is a sami install and the configuration file at root path.


PS 1. If I rember for example: "Stm_class"

**Comment by aik099** [comment]
There is no such `Stm_class` class in https://github.com/krzysiekpiasecki/Bowling repo. Or at least GitHub doesn't find it.

**Comment by krzysiekpiasecki** [comment]
I think you should reproduce bug with a repository above or wait for my detail answer to yours questions.

**Comment by krzysiekpiasecki** [comment]
1. string(10) "Stmt_Class"

2. https://github.com/krzysiekpiasecki/Bowling/blob/master/src/KrzysiekPiasecki/Bowling/BowlingGameInterface.php

3.  ```composer show -i```
```
blackfire/php-sdk                  v1.6.0  Blackfire.io PHP SDK
composer/ca-bundle                 1.0.6   Lets you find a path to the system CA bundle, and includes a...
doctrine/instantiator              1.0.5   A small, lightweight utility to instantiate objects in PHP w...
etsy/phan                          0.7     A static analyzer for PHP
friendsofphp/php-cs-fixer          v1.13.1 A tool to automatically fix PHP code style
michelf/php-markdown               1.7.0   PHP Markdown
myclabs/deep-copy                  1.5.5   Create deep copies (clones) of your objects
nikic/php-parser                   v1.4.1  A PHP parser written in PHP
phing/phing                        2.15.2  PHing Is Not GNU make; it's a PHP project build system or bu...
phpdocumentor/reflection-docblock  2.0.4   
phpspec/prophecy                   v1.6.2  Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage          4.0.3   Library that provides collection, processing, and rendering ...
phpunit/php-file-iterator          1.4.2   FilterIterator implementation that filters files based on a ...
phpunit/php-text-template          1.2.1   Simple template engine.
phpunit/php-timer                  1.0.8   Utility class for timing
phpunit/php-token-stream           1.4.9   Wrapper around PHP's tokenizer extension.
phpunit/phpunit                    5.7.3   The PHP Unit Testing framework.
phpunit/phpunit-mock-objects       3.4.3   Mock Object library for PHPUnit
pimple/pimple                      v3.0.2  Pimple, a simple Dependency Injection Container
psr/log                            1.0.2   Common interface for logging libraries
sami/sami                          v3.3.0  Sami, an API documentation generator
sebastian/code-unit-reverse-lookup 1.0.0   Looks up which function or method a line of code belongs to
sebastian/comparator               1.2.2   Provides the functionality to compare PHP values for equality
sebastian/diff                     1.4.1   Diff implementation
sebastian/environment              2.0.0   Provides functionality to handle HHVM/PHP environments
sebastian/exporter                 2.0.0   Provides the functionality to export PHP variables for visua...
sebastian/global-state             1.1.1   Snapshotting of global state
sebastian/object-enumerator        2.0.0   Traverses array structures and object graphs to enumerate al...
sebastian/recursion-context        2.0.0   Provides functionality to recursively process PHP variables
sebastian/resource-operations      1.0.0   Provides a list of PHP built-in functions that operate on re...
sebastian/version                  2.0.1   Library that helps with managing the version number of Git-h...
symfony/console                    v2.8.14 Symfony Console Component
symfony/debug                      v3.0.9  Symfony Debug Component
symfony/event-dispatcher           v3.2.0  Symfony EventDispatcher Component
symfony/filesystem                 v2.8.14 Symfony Filesystem Component
symfony/finder                     v2.8.14 Symfony Finder Component
symfony/polyfill-mbstring          v1.3.0  Symfony polyfill for the Mbstring extension
symfony/process                    v2.8.14 Symfony Process Component
symfony/stopwatch                  v3.2.0  Symfony Stopwatch Component
symfony/yaml                       v2.8.14 Symfony Yaml Component
twig/twig                          v1.28.2 Twig, the flexible, fast, and secure template language for PHP
```

It's a source of exception
class.twig

```twig
{% block class_signature -%}
    {% if class.interface and class.abstract %}abstract {% endif %}
    {{ class.categoryName|raw }}
    <strong>{{ class.shortname|raw }}</strong>
    {%- if class.parent %}
        extends {{ class_link(class.parent) }}
    {%- endif %}
    {%- if class.interfaces|length > 0 %}
        implements
        {% for interface in class.interfaces %}
            {{- class_link(interface) }}
            {%- if not loop.last %}, {% endif %}
        {%- endfor %}
    {%- endif %}
    {{- source_link(project, class) }}
{% endblock %}
```
 
Command output & exception trace string

```
string(10) "Stmt_Class"
  Parsing ### 7%                                               
  Parsing ####### 14%                                           
  Parsing ########## 21%                                        
  Parsing ############## 28%                                    
  Parsing ################# 35%                                 
  Parsing ##################### 42%                             
  Parsing ######################### 50%                         
  Parsing ############################ 57%                      
  Parsing ################################ 64%                  
  Parsing ################################### 71%               
  Parsing ####################################### 78%           
  Parsing ########################################## 85%        
  Parsing ############################################## 92%    
  Parsing   done
  Rendering ########################## 53%                        
            Class KrzysiekPiasecki\Bowling\BowlingGameInterface
string(251) "An exception has been thrown during the rendering of a template ("Warning: A non-numeric value encountered in /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Reflection/ClassReflection.php line 258") in "class.twig" at line 74."
string(5531) "#0 /Users/krzysztek/Documents/Workspaces/php/Bowling/var/cache/sami/twig/79/79a673f853f378a53ccdc7f57933946af88f01affdacb3050e5c294157cf1826.php(113): Twig_Template->displayBlock('class_signature', Array, Array)
#1 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(207): __TwigTemplate_011c46fa48eea978c774108b6ee76ff41fb21831c90df745a813ec482fca875b->block_page_content(Array, Array)
#2 /Users/krzysztek/Documents/Workspaces/php/Bowling/var/cache/sami/twig/4a/4adda73f924b0a79f907e7e24404803d01ea13a7de5d56e604dc2466dec90577.php(60): Twig_Template->displayBlock('page_content', Array, Array)
#3 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(207): __TwigTemplate_f4bb750546ecbb4e5d4c399bce406ff13bbf817086d82e57f68ccb774a2556b6->block_content(Array, Array)
#4 /Users/krzysztek/Documents/Workspaces/php/Bowling/var/cache/sami/twig/1a/1abad9b22e7bf0031f13b71c231c966ac4caa23581fc3ad7e77e29bec0229f5d.php(145): Twig_Template->displayBlock('content', Array, Array)
#5 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(207): __TwigTemplate_192f7be7e119ffb5f2ed78001359d974639e6856e4fb45df1793077fcd8b46c9->block_html(Array, Array)
#6 /Users/krzysztek/Documents/Workspaces/php/Bowling/var/cache/sami/twig/1a/1abad9b22e7bf0031f13b71c231c966ac4caa23581fc3ad7e77e29bec0229f5d.php(82): Twig_Template->displayBlock('html', Array, Array)
#7 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(428): __TwigTemplate_192f7be7e119ffb5f2ed78001359d974639e6856e4fb45df1793077fcd8b46c9->doDisplay(Array, Array)
#8 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(396): Twig_Template->displayWithErrorHandling(Array, Array)
#9 /Users/krzysztek/Documents/Workspaces/php/Bowling/var/cache/sami/twig/4a/4adda73f924b0a79f907e7e24404803d01ea13a7de5d56e604dc2466dec90577.php(30): Twig_Template->display(Array, Array)
#10 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(428): __TwigTemplate_f4bb750546ecbb4e5d4c399bce406ff13bbf817086d82e57f68ccb774a2556b6->doDisplay(Array, Array)
#11 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(396): Twig_Template->displayWithErrorHandling(Array, Array)
#12 /Users/krzysztek/Documents/Workspaces/php/Bowling/var/cache/sami/twig/79/79a673f853f378a53ccdc7f57933946af88f01affdacb3050e5c294157cf1826.php(43): Twig_Template->display(Array, Array)
#13 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(428): __TwigTemplate_011c46fa48eea978c774108b6ee76ff41fb21831c90df745a813ec482fca875b->doDisplay(Array, Array)
#14 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(396): Twig_Template->displayWithErrorHandling(Array, Array)
#15 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Template.php(407): Twig_Template->display(Array)
#16 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/twig/twig/lib/Twig/Environment.php(362): Twig_Template->render(Array)
#17 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Renderer/Renderer.php(197): Twig_Environment->render('class.twig', Array)
#18 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Renderer/Renderer.php(180): Sami\Renderer\Renderer->save(Object(Sami\Project), 'KrzysiekPiaseck...', 'class.twig', Array)
#19 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Renderer/Renderer.php(83): Sami\Renderer\Renderer->renderClassTemplates(Array, Object(Sami\Project), Array)
#20 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Project.php(493): Sami\Renderer\Renderer->render(Object(Sami\Project), Array, true)
#21 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Project.php(113): Sami\Project->renderVersion(Object(Sami\Version\Version), NULL, Array, true)
#22 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Console/Command/Command.php(79): Sami\Project->update(Array, true)
#23 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/Sami/Console/Command/UpdateCommand.php(52): Sami\Console\Command\Command->update(Object(Sami\Project))
#24 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/symfony/console/Command/Command.php(257): Sami\Console\Command\UpdateCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#25 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/symfony/console/Application.php(849): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#26 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/symfony/console/Application.php(193): Symfony\Component\Console\Application->doRunCommand(Object(Sami\Console\Command\UpdateCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#27 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/symfony/console/Application.php(124): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#28 /Users/krzysztek/Documents/Workspaces/php/Bowling/vendor/sami/sami/sami.php(14): Symfony\Component\Console\Application->run()
#29 {main}"

```


PS IMO It's a php 7.1 related. KP.

**Comment by aik099** [comment]
> nikic/php-parser                   v1.4.1  A PHP parser written in PHP

That's interesting. The 1.4.1 version of used PHP parser surely doesn't support that. Try `master` branch (not released yet, changes minimal required PHP version to 5.5 as well) of Sami that uses 3.x version of PHP parser.

**Comment by krzysiekpiasecki** [comment]
dev-master is broken ...

> [Twig_Error_Runtime]                                                 
  Block "search_index" on template "layout/base.twig" does not exist.

After changes and fixing base.twig the situation is still the same.

```composer```

```
blackfire/php-sdk                  v1.6.0             Blackfire.io PHP SDK
composer/ca-bundle                 dev-master a795611 Lets you find a path to the system CA bundle, and...
doctrine/instantiator              dev-master 416fb8a A small, lightweight utility to instantiate objec...
etsy/phan                          0.7                A static analyzer for PHP
friendsofphp/php-cs-fixer          1.13.x-dev 0ea4f7e A tool to automatically fix PHP code style
michelf/php-markdown               1.7.0              PHP Markdown
myclabs/deep-copy                  1.5.5              Create deep copies (clones) of your objects
nikic/php-parser                   dev-master 301c343 A PHP parser written in PHP
phing/phing                        dev-master a961a90 PHing Is Not GNU make; it's a PHP project build s...
phpdocumentor/reflection-docblock  2.0.4              
phpspec/prophecy                   dev-master 6c52c27 Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage          4.0.x-dev 31fbb24  Library that provides collection, processing, and...
phpunit/php-file-iterator          dev-master 3cc8f69 FilterIterator implementation that filters files ...
phpunit/php-text-template          1.2.1              Simple template engine.
phpunit/php-timer                  1.0.8              Utility class for timing
phpunit/php-token-stream           dev-master 3b402f6 Wrapper around PHP's tokenizer extension.
phpunit/phpunit                    5.7.x-dev 94cec03  The PHP Unit Testing framework.
phpunit/phpunit-mock-objects       3.4.3              Mock Object library for PHPUnit
pimple/pimple                      dev-master 1ce4ec4 Pimple, a simple Dependency Injection Container
psr/log                            dev-master 4ebe3a8 Common interface for logging libraries
sami/sami                          dev-master 2048fcb Sami, an API documentation generator
sebastian/code-unit-reverse-lookup dev-master 9980576 Looks up which function or method a line of code ...
sebastian/comparator               dev-master 2f09d5a Provides the functionality to compare PHP values ...
sebastian/diff                     dev-master d081431 Diff implementation
sebastian/environment              dev-master 5795ffe Provides functionality to handle HHVM/PHP environ...
sebastian/exporter                 dev-master ce474bd Provides the functionality to export PHP variable...
sebastian/global-state             dev-master ab3e5ce Snapshotting of global state
sebastian/object-enumerator        dev-master 96f8a3f Traverses array structures and object graphs to e...
sebastian/recursion-context        dev-master 2c3ba15 Provides functionality to recursively process PHP...
sebastian/resource-operations      dev-master fadc83f Provides a list of PHP built-in functions that op...
sebastian/version                  dev-master 99732be Library that helps with managing the version numb...
symfony/console                    dev-master a7a4ffc Symfony Console Component
symfony/debug                      dev-master 342af93 Symfony Debug Component
symfony/event-dispatcher           dev-master b0cabf5 Symfony EventDispatcher Component
symfony/filesystem                 dev-master 696d630 Symfony Filesystem Component
symfony/finder                     dev-master 941debb Symfony Finder Component
symfony/polyfill-mbstring          dev-master e79d363 Symfony polyfill for the Mbstring extension
symfony/process                    dev-master b2a461e Symfony Process Component
symfony/stopwatch                  dev-master a538e8a Symfony Stopwatch Component
symfony/yaml                       dev-master f8c054e Symfony Yaml Component
twig/twig                          2.x-dev 5c0aabb    Twig, the flexible, fast, and secure template lan...
```


I think it's only php 7.1 related and it's a bug - binary operation with string argument.

**Comment by aik099** [comment]
Could you please send separate PRs that fixes:

1. Block "search_index" on template "layout/base.twig" does not exist.
2. binary operation with string argument.

**Comment by aik099** [comment]
Since `$this->modifiers` is `Stmt_Class` that might as well can be problem in https://github.com/nikic/PHP-Parser.

**Comment by krzysiekpiasecki** [comment]
It's only php 7.1 related and it's a hidden old bug - binary operation with string argument.
- http://php.net/manual/en/migration71.other-changes.php

**Comment by krzysiekpiasecki** [comment]
@aik099 Sorry but I can't handle. Thanks for participating in this issue.

**Comment by aik099** [comment]
Not sure why you're closing and reopening this issue several times. Probably you need to keep issue opened, because problem isn't solved.

**Comment by krzysiekpiasecki** [comment]
It's hard to explain. Sorry about that.

**Comment by fabpot** [comment]
fixed in d145117
